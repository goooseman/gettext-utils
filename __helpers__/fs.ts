import * as fse from "fs-extra";
import * as os from "os";
import * as path from "path";

export const readFileFromTmp = (tmpPath: string, filePath: string) => {
  return fse.readFileSync(path.join(tmpPath, filePath)).toString();
};

export const getTmpPath = async () => {
  const exportFolder = path.join(os.tmpdir(), `lioness-utils`);
  await fse.remove(exportFolder);
  return exportFolder;
};
