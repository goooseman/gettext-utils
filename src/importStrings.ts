import * as fse from "fs-extra";
import { po } from "gettext-parser";
import * as glob from "glob";
import * as path from "path";

import { optimizeAllTranslations } from "./utils/optimizeTranslations";

export const getPoParsed = async (pathToPo: string) => {
  const file = await fse.readFile(pathToPo);
  return po.parse(file.toString());
};

export const getTranslations = async (poFilesPath: string) => {
  const poFilesPaths = glob.sync(path.join(poFilesPath, "*.po"));
  const translations: TranslationsMap = {};
  for (const poFilePath of poFilesPaths) {
    const locale = path.basename(poFilePath, path.extname(poFilePath));
    translations[locale] = await getPoParsed(poFilePath);
  }
  return translations;
};

const writeTranslationsByLocale = async (
  translations: TranslationsMap,
  outputFilePath: string,
) => {
  const extname = path.extname(outputFilePath).replace(".", "");
  // If file path was passed instead of a folder, then remove it's
  // extension, create a folder by its path and put translations there
  const localeOutputPath = extname
    ? outputFilePath.replace(new RegExp(`(\.${extname})$`, "gi"), "")
    : outputFilePath;
  await fse.mkdirp(localeOutputPath);

  for (const locale of Object.keys(translations)) {
    await fse.writeFile(
      path.join(localeOutputPath, `./${locale}.json`),
      JSON.stringify(translations[locale]),
    );
  }
};

const importStrings = async (
  poFilesDirPath: string,
  outputFilePath: string,
  optimize?: boolean,
  splitByLocale?: boolean,
) => {
  const translations = await getTranslations(poFilesDirPath);
  const finalTranslations = optimize
    ? optimizeAllTranslations(translations)
    : translations;
  await fse.mkdirp(path.dirname(outputFilePath));

  if (splitByLocale) {
    await writeTranslationsByLocale(finalTranslations, outputFilePath);
  } else {
    await fse.writeFile(outputFilePath, JSON.stringify(finalTranslations));
  }
};

export default importStrings;
