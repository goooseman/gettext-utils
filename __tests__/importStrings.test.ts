import importStrings from "@src/importStrings";
import { getTmpPath } from "__helpers__/fs";
import * as fse from "fs-extra";
import * as path from "path";

it("should generate node-gettext .json files from .po files", async () => {
  const tmpPath = await getTmpPath();
  const filePath = path.join(tmpPath, "translations.json");
  await importStrings("__fixtures__/po", filePath);

  const file = JSON.parse(fse.readFileSync(filePath, "utf-8"));
  expect(file).toMatchSnapshot();
  const localesToCheck = ["he", "fr", "ru"];
  for (const localeToCheck of localesToCheck) {
    expect(file).toHaveProperty(localeToCheck);
  }
});

it("can optimize translations if boolean flag is passed", async () => {
  const tmpPath = await getTmpPath();
  const filePath = path.join(tmpPath, "translations.json");
  await importStrings("__fixtures__/po", filePath, true);

  const file = JSON.parse(
    fse.readFileSync(filePath, "utf-8"),
  ) as TranslationsMap;
  expect(file).toMatchSnapshot();

  const translation = file.en;
  expect(translation).not.toHaveProperty("charset");
  expect(Object.keys(translation.headers)).toHaveLength(1);
  expect(translation.headers).toHaveProperty("plural-forms");

  const message = translation.translations["lion.females"]["one female lion"];
  expect(message).not.toHaveProperty("msgid");
  expect(message).not.toHaveProperty("comments");
});

it("allows splitting translations by locales", async () => {
  const tmpPath = await getTmpPath();
  const translationsPath = path.join(tmpPath, "translations");
  await importStrings("__fixtures__/po", translationsPath, false, true);

  const enJson = JSON.parse(
    fse.readFileSync(path.join(translationsPath, "./en.json"), "utf-8"),
  ) as Translation;
  const heJson = JSON.parse(
    fse.readFileSync(path.join(translationsPath, "./he.json"), "utf-8"),
  ) as Translation;
  const ruJson = JSON.parse(
    fse.readFileSync(path.join(translationsPath, "./ru.json"), "utf-8"),
  ) as Translation;

  expect(enJson).toMatchSnapshot();
  expect(heJson).toMatchSnapshot();
  expect(ruJson).toMatchSnapshot();
  expect(enJson.translations).toBeTruthy();
  expect(heJson.translations).toBeTruthy();
  expect(ruJson.translations).toBeTruthy();
});
