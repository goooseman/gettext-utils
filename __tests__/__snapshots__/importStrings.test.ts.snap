// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows splitting translations by locales 1`] = `
Object {
  "charset": "utf-8",
  "headers": Object {
    "content-transfer-encoding": "8bit",
    "content-type": "text/plain; charset=UTF-8",
    "language": "en",
    "language-team": "",
    "last-translator": "",
    "mime-version": "1.0",
    "plural-forms": "nplurals=2; plural=(n != 1);",
    "po-revision-date": "",
    "pot-creation-date": "Thu Jan 01 2018 00:00:00",
    "project-id-version": "",
    "x-generator": "Poedit 2.2.1",
  },
  "translations": Object {
    "": Object {
      "": Object {
        "msgid": "",
        "msgstr": Array [
          "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=2; plural=(n != 1);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: en
",
        ],
      },
    },
    "lion.females": Object {
      "one female lion": Object {
        "comments": Object {
          "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:23",
        },
        "msgctxt": "lion.females",
        "msgid": "one female lion",
        "msgid_plural": "{{ count }} female lions",
        "msgstr": Array [
          "one female lion",
          "{{ count }} female lions",
        ],
      },
    },
    "lion.sound": Object {
      "Rrrrr": Object {
        "comments": Object {
          "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:19",
        },
        "msgctxt": "lion.sound",
        "msgid": "Rrrrr",
        "msgstr": Array [
          "Rrrr",
        ],
      },
    },
  },
}
`;

exports[`allows splitting translations by locales 2`] = `
Object {
  "charset": "utf-8",
  "headers": Object {
    "content-transfer-encoding": "8bit",
    "content-type": "text/plain; charset=UTF-8",
    "language": "he",
    "language-team": "",
    "last-translator": "",
    "mime-version": "1.0",
    "plural-forms": "nplurals=4; plural=(n==1 ? 0 : n==2 ? 1 : n>10 && n%10==0 ? 2 : 3);",
    "po-revision-date": "",
    "pot-creation-date": "Thu Jan 01 2018 00:00:00",
    "project-id-version": "",
    "x-generator": "Poedit 2.2.1",
  },
  "translations": Object {
    "": Object {
      "": Object {
        "msgid": "",
        "msgstr": Array [
          "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=4; plural=(n==1 ? 0 : n==2 ? 1 : n>10 && n%10==0 ? 2 : 3);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: he
",
        ],
      },
    },
    "lion.females": Object {
      "one female lion": Object {
        "comments": Object {
          "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:23",
        },
        "msgctxt": "lion.females",
        "msgid": "one female lion",
        "msgid_plural": "{{ count }} female lions",
        "msgstr": Array [
          "לביאה",
          "שתי לביאה",
          "‫{{count}} לביאות",
          "‏‫{{count}} לביא",
        ],
      },
    },
    "lion.sound": Object {
      "Rrrrr": Object {
        "comments": Object {
          "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:19",
        },
        "msgctxt": "lion.sound",
        "msgid": "Rrrrr",
        "msgstr": Array [
          "רררר",
        ],
      },
    },
  },
}
`;

exports[`allows splitting translations by locales 3`] = `
Object {
  "charset": "utf-8",
  "headers": Object {
    "content-transfer-encoding": "8bit",
    "content-type": "text/plain; charset=UTF-8",
    "language": "ru",
    "language-team": "",
    "last-translator": "",
    "mime-version": "1.0",
    "plural-forms": "nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<12 || n%100>14) ? 1 : 2);",
    "po-revision-date": "",
    "pot-creation-date": "Thu Jan 01 2018 00:00:00",
    "project-id-version": "",
    "x-generator": "Poedit 2.2.1",
  },
  "translations": Object {
    "": Object {
      "": Object {
        "msgid": "",
        "msgstr": Array [
          "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<12 || n%100>14) ? 1 : 2);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: ru
",
        ],
      },
    },
    "lion.females": Object {
      "one female lion": Object {
        "comments": Object {
          "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:23",
        },
        "msgctxt": "lion.females",
        "msgid": "one female lion",
        "msgid_plural": "{{ count }} female lions",
        "msgstr": Array [
          "одна львица",
          "{{ сount }} львицы",
          "{{ count }} львиц",
        ],
      },
    },
    "lion.sound": Object {
      "Rrrrr": Object {
        "comments": Object {
          "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:19",
        },
        "msgctxt": "lion.sound",
        "msgid": "Rrrrr",
        "msgstr": Array [
          "Ррррр",
        ],
      },
    },
  },
}
`;

exports[`can optimize translations if boolean flag is passed 1`] = `
Object {
  "en": Object {
    "headers": Object {
      "plural-forms": "nplurals=2; plural=(n != 1);",
    },
    "translations": Object {
      "": Object {
        "": Object {
          "msgstr": Array [
            "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=2; plural=(n != 1);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: en
",
          ],
        },
      },
      "lion.females": Object {
        "one female lion": Object {
          "msgctxt": "lion.females",
          "msgid_plural": "{{ count }} female lions",
          "msgstr": Array [
            "one female lion",
            "{{ count }} female lions",
          ],
        },
      },
      "lion.sound": Object {
        "Rrrrr": Object {
          "msgctxt": "lion.sound",
          "msgstr": Array [
            "Rrrr",
          ],
        },
      },
    },
  },
  "fr": Object {
    "headers": Object {
      "plural-forms": "nplurals=2; plural=(n > 1);",
    },
    "translations": Object {
      "": Object {
        "": Object {
          "msgstr": Array [
            "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=2; plural=(n > 1);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: fr
",
          ],
        },
      },
      "fake": Object {
        "should be deleted": Object {
          "msgctxt": "fake",
          "msgid_plural": "Delete me {{ count }} times",
          "msgstr": Array [
            "Delete me",
            "Delete me {{ count }} times",
          ],
        },
      },
      "lion.females": Object {
        "one female lion": Object {
          "msgctxt": "lion.females",
          "msgid_plural": "{{ count }} female lions",
          "msgstr": Array [
            "lionne",
            "{{ count }} lionnes",
          ],
        },
      },
      "lion.sound": Object {
        "Rrrrr": Object {
          "msgctxt": "lion.sound",
          "msgstr": Array [
            "Oooo",
          ],
        },
      },
    },
  },
  "he": Object {
    "headers": Object {
      "plural-forms": "nplurals=4; plural=(n==1 ? 0 : n==2 ? 1 : n>10 && n%10==0 ? 2 : 3);",
    },
    "translations": Object {
      "": Object {
        "": Object {
          "msgstr": Array [
            "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=4; plural=(n==1 ? 0 : n==2 ? 1 : n>10 && n%10==0 ? 2 : 3);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: he
",
          ],
        },
      },
      "lion.females": Object {
        "one female lion": Object {
          "msgctxt": "lion.females",
          "msgid_plural": "{{ count }} female lions",
          "msgstr": Array [
            "לביאה",
            "שתי לביאה",
            "‫{{count}} לביאות",
            "‏‫{{count}} לביא",
          ],
        },
      },
      "lion.sound": Object {
        "Rrrrr": Object {
          "msgctxt": "lion.sound",
          "msgstr": Array [
            "רררר",
          ],
        },
      },
    },
  },
  "ru": Object {
    "headers": Object {
      "plural-forms": "nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<12 || n%100>14) ? 1 : 2);",
    },
    "translations": Object {
      "": Object {
        "": Object {
          "msgstr": Array [
            "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<12 || n%100>14) ? 1 : 2);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: ru
",
          ],
        },
      },
      "lion.females": Object {
        "one female lion": Object {
          "msgctxt": "lion.females",
          "msgid_plural": "{{ count }} female lions",
          "msgstr": Array [
            "одна львица",
            "{{ сount }} львицы",
            "{{ count }} львиц",
          ],
        },
      },
      "lion.sound": Object {
        "Rrrrr": Object {
          "msgctxt": "lion.sound",
          "msgstr": Array [
            "Ррррр",
          ],
        },
      },
    },
  },
}
`;

exports[`should generate node-gettext .json files from .po files 1`] = `
Object {
  "en": Object {
    "charset": "utf-8",
    "headers": Object {
      "content-transfer-encoding": "8bit",
      "content-type": "text/plain; charset=UTF-8",
      "language": "en",
      "language-team": "",
      "last-translator": "",
      "mime-version": "1.0",
      "plural-forms": "nplurals=2; plural=(n != 1);",
      "po-revision-date": "",
      "pot-creation-date": "Thu Jan 01 2018 00:00:00",
      "project-id-version": "",
      "x-generator": "Poedit 2.2.1",
    },
    "translations": Object {
      "": Object {
        "": Object {
          "msgid": "",
          "msgstr": Array [
            "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=2; plural=(n != 1);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: en
",
          ],
        },
      },
      "lion.females": Object {
        "one female lion": Object {
          "comments": Object {
            "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:23",
          },
          "msgctxt": "lion.females",
          "msgid": "one female lion",
          "msgid_plural": "{{ count }} female lions",
          "msgstr": Array [
            "one female lion",
            "{{ count }} female lions",
          ],
        },
      },
      "lion.sound": Object {
        "Rrrrr": Object {
          "comments": Object {
            "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:19",
          },
          "msgctxt": "lion.sound",
          "msgid": "Rrrrr",
          "msgstr": Array [
            "Rrrr",
          ],
        },
      },
    },
  },
  "fr": Object {
    "charset": "utf-8",
    "headers": Object {
      "content-transfer-encoding": "8bit",
      "content-type": "text/plain; charset=UTF-8",
      "language": "fr",
      "language-team": "",
      "last-translator": "",
      "mime-version": "1.0",
      "plural-forms": "nplurals=2; plural=(n > 1);",
      "po-revision-date": "",
      "pot-creation-date": "Thu Jan 01 2018 00:00:00",
      "project-id-version": "",
      "x-generator": "Poedit 2.2.1",
    },
    "translations": Object {
      "": Object {
        "": Object {
          "msgid": "",
          "msgstr": Array [
            "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=2; plural=(n > 1);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: fr
",
          ],
        },
      },
      "fake": Object {
        "should be deleted": Object {
          "comments": Object {
            "reference": "some component that does not exists",
          },
          "msgctxt": "fake",
          "msgid": "should be deleted",
          "msgid_plural": "Delete me {{ count }} times",
          "msgstr": Array [
            "Delete me",
            "Delete me {{ count }} times",
          ],
        },
      },
      "lion.females": Object {
        "one female lion": Object {
          "comments": Object {
            "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:23",
          },
          "msgctxt": "lion.females",
          "msgid": "one female lion",
          "msgid_plural": "{{ count }} female lions",
          "msgstr": Array [
            "lionne",
            "{{ count }} lionnes",
          ],
        },
      },
      "lion.sound": Object {
        "Rrrrr": Object {
          "comments": Object {
            "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:19",
          },
          "msgctxt": "lion.sound",
          "msgid": "Rrrrr",
          "msgstr": Array [
            "Oooo",
          ],
        },
      },
    },
  },
  "he": Object {
    "charset": "utf-8",
    "headers": Object {
      "content-transfer-encoding": "8bit",
      "content-type": "text/plain; charset=UTF-8",
      "language": "he",
      "language-team": "",
      "last-translator": "",
      "mime-version": "1.0",
      "plural-forms": "nplurals=4; plural=(n==1 ? 0 : n==2 ? 1 : n>10 && n%10==0 ? 2 : 3);",
      "po-revision-date": "",
      "pot-creation-date": "Thu Jan 01 2018 00:00:00",
      "project-id-version": "",
      "x-generator": "Poedit 2.2.1",
    },
    "translations": Object {
      "": Object {
        "": Object {
          "msgid": "",
          "msgstr": Array [
            "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=4; plural=(n==1 ? 0 : n==2 ? 1 : n>10 && n%10==0 ? 2 : 3);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: he
",
          ],
        },
      },
      "lion.females": Object {
        "one female lion": Object {
          "comments": Object {
            "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:23",
          },
          "msgctxt": "lion.females",
          "msgid": "one female lion",
          "msgid_plural": "{{ count }} female lions",
          "msgstr": Array [
            "לביאה",
            "שתי לביאה",
            "‫{{count}} לביאות",
            "‏‫{{count}} לביא",
          ],
        },
      },
      "lion.sound": Object {
        "Rrrrr": Object {
          "comments": Object {
            "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:19",
          },
          "msgctxt": "lion.sound",
          "msgid": "Rrrrr",
          "msgstr": Array [
            "רררר",
          ],
        },
      },
    },
  },
  "ru": Object {
    "charset": "utf-8",
    "headers": Object {
      "content-transfer-encoding": "8bit",
      "content-type": "text/plain; charset=UTF-8",
      "language": "ru",
      "language-team": "",
      "last-translator": "",
      "mime-version": "1.0",
      "plural-forms": "nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<12 || n%100>14) ? 1 : 2);",
      "po-revision-date": "",
      "pot-creation-date": "Thu Jan 01 2018 00:00:00",
      "project-id-version": "",
      "x-generator": "Poedit 2.2.1",
    },
    "translations": Object {
      "": Object {
        "": Object {
          "msgid": "",
          "msgstr": Array [
            "Content-Type: text/plain; charset=UTF-8
POT-Creation-Date: Thu Jan 01 2018 00:00:00
Content-Transfer-Encoding: 8bit
Plural-Forms: nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<12 || n%100>14) ? 1 : 2);
Project-Id-Version: 
PO-Revision-Date: 
Language-Team: 
MIME-Version: 1.0
X-Generator: Poedit 2.2.1
Last-Translator: 
Language: ru
",
          ],
        },
      },
      "lion.females": Object {
        "one female lion": Object {
          "comments": Object {
            "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:23",
          },
          "msgctxt": "lion.females",
          "msgid": "one female lion",
          "msgid_plural": "{{ count }} female lions",
          "msgstr": Array [
            "одна львица",
            "{{ сount }} львицы",
            "{{ count }} львиц",
          ],
        },
      },
      "lion.sound": Object {
        "Rrrrr": Object {
          "comments": Object {
            "reference": "__fixtures__/react-project/src/components/LionessHocComponent.tsx:19",
          },
          "msgctxt": "lion.sound",
          "msgid": "Rrrrr",
          "msgstr": Array [
            "Ррррр",
          ],
        },
      },
    },
  },
}
`;
